### ISSUE TITLE
gcsfuse becomes very slow when accessed by multiple threads/processes

### ISSUE THREAD

**Issue by nmousavi** [GoogleCloudPlatform/gcsfuse#262]
Experimenting with using gcsfuse for DeepVariant (https://github.com/google/deepvariant), I've noticed a big performance degradation when more than one instance is reading the data via gcsfuse (read only). 

Observations:

- Input data is ~70GB.
- A single instance processes 1/2048 of the data in 15min when input data is downloaded from gcs.
- A single instance processes 1/2048 of the data in 12min when input data is fused using gcsfuse.
- 8 instances (running in parallel in a GCE VM) process 8/2048 of the data in 16min when input data is downloaded from gcs.
- 8 instances (running in parallel in a GCE VM) process 8/2048 of the data in ~30hours when input data is fused using gcsfuse.

Access pattern is interleaved, i.e., 

process 1 reads bytes 1-1000, 2048*1000 - 2048*1000 + 1000, so on
process 2 reads bytes 2000-3000, 2048*1000 + 2000 - 2048* 1000 + 3000, son 
.....

Any idea why the performance is so poor in multiprocess mode?  Please let me know if debug info needed. Thanks!

**Comment by jacobsa** [comment]
To confirm, are you running all of these processes on a single VM? Have you considered running them on separate VMs (this would fully isolate them), or with separate gcsfuse mount points? I expect you'd see the performance problem entirely go away with either of these.

I imagine the bottleneck here is either a lock in the kernel (you could probably see this with `gcsfuse --foreground --debug_fuse`) or maybe a lock in gcsfuse (we might be able to figure this out by looking at `gcsfuse --foreground --debug_fuse --debug_gcs`).

**Comment by nmousavi** [comment]
Yes, we used a single VM. Running them on separate VMs is not an scalable option for us, but running separate gcsfuse is an idea we will try (and update you). Is there any limit on the number of separate mount points we can/should create?

I am still interested in knowing why this happens in multi-process scenario, so collected the logs here:

https://paste.googleplex.com/6560735928778752
https://paste.googleplex.com/6478377296855040

Related question: why is `limit-ops-per-sec` set to 5? I have changed it to no limit (-1), but haven't seen any improvement .

**Comment by jacobsa** [comment]
There's no limit on the number of mount points. `--limit-ops-per-sec` is set to a reasonable value by default just to avoid surprise huge bills from GCS.

Thanks for the log files, but there's an awful lot going on there. (I also don't see `--debug_gcs` outpout?) If you're able to narrow this down a bit more to a small test case that reproduces this, or annotate the logs with what you're seeing etc, I could take a look.

(gcsfuse should probably also be made to include timestamps in the log by changing the flags to `log.New` [here](https://github.com/GoogleCloudPlatform/gcsfuse/blob/957d9eae7eea4a1bf3161991337044713ad09d58/mount.go#L132) and [here](https://github.com/GoogleCloudPlatform/gcsfuse/blob/957d9eae7eea4a1bf3161991337044713ad09d58/main.go#L216). Unfortunately I can't get to this at the moment.)

**Comment by nmousavi** [comment]
Running separate gcsfuse solved the issue. To confirm, we shouldn't be worried if the number of gcsfuse instances grow to 64 for example? Also, what is the status of this project? We are planning to use it in prod. I would be happy to chat with you why we think gcsfuse can have significant impact in reducing the costs for our usecase.

re logs: Not sure why debug_gcs output is not there, I have set both flags on. I can redo it on a smaller test anyway (and will update you). Regarding timestamp, I think it will be really a good feature to have. I may look into it.

**Comment by nmousavi** [comment]
Sent PR #263 for including timestamp in debug los.

**Comment by nmousavi** [comment]
Log for a smaller (x100) test case:

https://nmousavi.users.x20web.corp.google.com/gcsfuse.log

I've noticed the download time for 1MB read can be as large as 5sec.

**Comment by jacobsa** [comment]
Thanks. I can see from the log that we're receiving interleaved requests for the same inode (but different handles) from the kernel, so this isn't a kernel lock. But it appears that we only ever start a GCS read after the previous one has returned, so it seems like this is gcsfuse-internal locking.

[Here's](https://github.com/GoogleCloudPlatform/gcsfuse/blob/4abce35dd15ea07f0903c662d3ebf9f531c05cd4/internal/fs/fs.go#L1542-L1563) the implementation of ReadFile, which calls [FileHandle.Read](https://github.com/GoogleCloudPlatform/gcsfuse/blob/4abce35dd15ea07f0903c662d3ebf9f531c05cd4/internal/fs/handle/file.go#L77-L122). That function has code that attempts to unlock the inode and allow parallel reads to proceed, but I guess maybe it's getting defeated? Adding some logging there and in [FileHandle.tryEnsureRead](https://github.com/GoogleCloudPlatform/gcsfuse/blob/4abce35dd15ea07f0903c662d3ebf9f531c05cd4/internal/fs/handle/file.go#L136-L173) may shed light on why this is.

**Comment by jacobsa** [comment]
Note that I can't tell for sure from that log whether the reads are being serialized. Adding logging in those functions would help answer that.

**Comment by dalexander** [comment]
+1, also interested in understanding this

**Comment by nmousavi** [comment]
I have added some more logging and the result is here (~45MB)

http://shortn/_nF4y59uokV

Logging CL here:
https://github.com/GoogleCloudPlatform/gcsfuse/compare/master...nmousavi:more-logging?expand=1

I don't see anything that could explain the issue. Any suggestion for additional logging?

**Comment by jacobsa** [comment]
Thanks very much for that.

I noticed a few things. (By the way, I see you're hosting these logs non-publicly; do you mind if I post excerpts?)

1. The reads within a file handle aren't contiguous. For example, handle 16 reads [3002368, 3133440), and then next reads at offset 5623808. Is this expected? I assume it's because you're using `pread` to do a random read of the next available piece of work. But unfortunately it defeats [the optimization](https://github.com/GoogleCloudPlatform/gcsfuse/blob/25c2c1b11b9dbc34920cbf5a4da236ab2d9a0341/internal/gcsx/random_reader.go#L114-L129) in gcsx.randomReader.ReadAt of reusing an existing read. Is it possible for you to consume sequentially from all of your file descriptors?

2. You can see from looking at just the outgoing GCS requests that you're reading the same data from GCS many times, wasting bandwidth and CPU. For example, in one request you read [5361664, 6410240), and then in the next one you read [5492736, 6541312), which is slid forward only 128 KiB despite the reads being 1 MiB. This is probably a very large source of wastage, and it's due to defeating the optimization I mentioned above.

3. It's pretty clear from the log that we're not blocking on file system or inode locks. Good.

4. But it's also clear from using `grep gcs:` that there are stalls in the GCS requests. We'll see no read to GCS issued for maybe 100 ms, and then as soon as one GCS response comes back we'll issue the next one. From the logging you added there seems to be some rate- or in-flight limit after the [call](https://github.com/GoogleCloudPlatform/gcsfuse/blob/25c2c1b11b9dbc34920cbf5a4da236ab2d9a0341/internal/gcsx/random_reader.go#L271-L280) to Bucket.NewReader in gcsx.randomReader.startRad.

5. To confirm, are you positive you're not using any rate limiting? The wait [here](https://github.com/GoogleCloudPlatform/gcsfuse/blob/284558bd039f4859a5fd4b465639fe69725e34a0/vendor/github.com/jacobsa/ratelimit/throttled_bucket.go#L54-L76) would explain this. Maybe add some logging to confirm?

6. The only other place I could think where you might be blocking is in the HTTP library itself, perhaps waiting on availability of HTTP/2 flow control tokens or similar. Maybe add logging within [gcs.bucket.NewReader](https://github.com/GoogleCloudPlatform/gcsfuse/blob/284558bd039f4859a5fd4b465639fe69725e34a0/vendor/github.com/jacobsa/gcloud/gcs/read.go#L31-L134) or `net/http` itself to confirm?

**Comment by nmousavi** [comment]
0. Sure, DeepVariant is open source after all.
1. Yes, this is expected. The intuition behind this is that we don't know the computational work required for each peace of input data, so 1MB interleaving is an attempt to assign fair amount of work to each worker.

2. This is somewhat strange to me, I have asked other people about this internally. However, it's not  causing a real performance issue for us (running multiple gcsfuse gives a satisfactory result).

3. Yes, it's clear from the log. Good to see it.

4, 5. Yes, we don't have any throttling neither in gcsfuse nor in our side (multiple gcsfuse scenario works just fine). I've added logging to `opThrottle.Wait` (see log and PR below), but nothing for this call.

6. Added time logging for `bucket.Reader` as well, but apparently http req handling is quick too (~100ms). Search for `http response after` in the log.

Log (~25MB):
http://shortn/_hUlqu1TqEE

PR:
https://github.com/GoogleCloudPlatform/gcsfuse/compare/master...nmousavi:more-logging?expand=1

**Comment by jacobsa** [comment]
I’ll try to look at the log on Monday. In the meantime, it occurs to me
that if you were to send SIGQUIT to gcsfuse at any random time, you might
see whatever is causing the blocking in the backtrace printed.

**Comment by jacobsa** [comment]
A couple more thoughts:

1. Instead of having N file descriptors for N workers, can you have one file descriptor that you read the entire file sequentially from, and then hand out pieces of the data read to the workers?

2. Assuming not, can you make your read sizes match the 1 MiB initial read size made by gcsfuse?

3. Assuming not, it may make sense to make the 1 MiB magic number for min read size configurable. In this case you'd want to set it to 1 byte (i.e. no read blowups).

**Comment by jacobsa** [comment]
(You also need to fix the permissions on the log.)

**Comment by nmousavi** [comment]
1. Not desired as it requires code change for the binary we don't own. 
2. Read size is determine with some third party lib and depends on the input data. I doubt we can change it. But let me understand the situation first. What happen if the read size is less than minReadSize=1MB, say 4k? Assume next read requests is >1MB apart from the end of prv read req? Does gcsfuse bundles these two read requests, and send a req to GCS for a region that contains both 4k reads? How the suggestion in 3 may help the situation?

3. I'll try this. 

(Access granted)

**Comment by nmousavi** [comment]
Looking into system call logs, I see `DeepVariant` (`DV`) makes a very large (exactly 128MB!?!) read request to `gcsfuse` every so often, and this is when everything freezes for minutes (1-3min); `DV` waits for read() sys call, and gcsfuse waits for some lock.

<pre>
strace log for DV <br>
19:06:23 lseek(6, 18099034291, SEEK_SET) = 18099034291
<b>19:06:23 </b>read(6"\37\213\10\4\0\0\0\0\0\377\6\0BC\2\0_L\315\275\t\230$\371U\37\3303\243\356\254\256"..., <b>134217728</b>) = <b>134217728</b>
<b>19:08:24</b> lseek(4, 235593595, SEEK_SET)  = 235593595
19:08:24 read(4, "\37\213\10\4\0\0\0\0\0\377\6\0BC\2\0uM\235\235[\226\344J\216k\3775+.\373\340"..., 4096) = 4096
</pre>

<pre>
strace log for gcsfuse <br>
19:06:25 futex(0xc820027508, FUTEX_WAKE, 1) = 1
19:06:25 futex(0xc75128, FUTEX_WAIT, 0, NULL) = 0 
19:06:26 futex(0xc75128, FUTEX_WAIT, 0, NULL) = 0
<b>19:06:27</b> futex(0xc75128, FUTEX_WAIT, 0, NULL) = 0 
<b>19:07:30</b> futex(0xc820027908, FUTEX_WAKE, 1) = 1     <b># 1 min lock wait!</b>
19:07:30 read(7, "\2533d8\231@,\22<#\1\332\231\31\207\325\333x\372k\331d\26\337t'\f\373y\207V\30"..., 8230) = 8230
19:07:30 sched_yield()                  = 0

</pre>


We can see gcsfuse spends most of its time (13%) on kernel lock `__lock_text_start`.

<img width="1440" alt="screen shot 2018-02-28 at 10 31 00 pm" src="https://user-images.githubusercontent.com/4221028/36825789-5123c33a-1cd7-11e8-8b10-83440469c1d5.png">

From Network dump, I see more Application Data packet and relatively slow Acks are sent to `GCS` during this time. For example, only 454MB was acked between [19:36:07, 19:37:22] which is only 7MB/s. No TCP zeroWindow Err in this particular period, but this is something that happens for `gcsfuse` (not very frequent though).

```
11	19:36:07.460602	10.142.0.8	74.125.141.95	TCP	66	34908 → 443 [ACK] Seq=67 Ack=40150 Win=24323 Len=0 TSval=609828941 TSecr=3815364206
...
83526	19:37:22.274651	10.142.0.8	74.125.141.95	TCP	66	34908 → 443 [ACK] Seq=75858 Ack=476136771 Win=24020 Len=0 TSval=609903749 TSecr=3815439013
```

TCP zero window

<img width="1440" alt="screen shot 2018-02-28 at 11 01 59 am" src="https://user-images.githubusercontent.com/4221028/36848116-4e5c9634-1d2e-11e8-888a-63f1c7ac4807.png">

I sent a `SIGQUIT` to `gcsfuse` when it was stalled on the lock:

There are many awaitRequestCancel thread with O(min) wait

```
goroutine 91265 [select, 4 minutes]:
net/http.(*http2clientStream).awaitRequestCancel(0xc8e1b9c3c0, 0xc8223e68a0)
	/usr/lib/go-1.6/src/net/http/h2_bundle.go:4731 +0x156
created by net/http.(*http2clientConnReadLoop).processHeaderBlockFragment
	/usr/lib/go-1.6/src/net/http/h2_bundle.go:5731 +0x7c9

goroutine 92709 [select, 4 minutes]:
net/http.(*http2clientStream).awaitRequestCancel(0xc84680d040, 0xc9bc313da0)
	/usr/lib/go-1.6/src/net/http/h2_bundle.go:4731 +0x156
created by net/http.(*http2clientConnReadLoop).processHeaderBlockFragment
	/usr/lib/go-1.6/src/net/http/h2_bundle.go:5731 +0x7c9
...
```


[1] network dump [here](https://nmousavi.users.x20web.corp.google.com/tcpdump-gcsfuse.pcap) (~1GB)
[2] `gcsfuse` log with `SIGQUIT` [here](https://nmousavi.users.x20web.corp.google.com/gcsfuse-sigquit.log) (36MB)

**Comment by nmousavi** [comment]
(Also changing [minReadSize](https://github.com/GoogleCloudPlatform/gcsfuse/blob/25c2c1b11b9dbc34920cbf5a4da236ab2d9a0341/internal/gcsx/random_reader.go#L26-L28) to 1B doesn't help)

**Comment by nmousavi** [comment]
Figured out that `DV` reads from two files: 1) BAM file (fd=6) always in 128MB chunks, 2) FASTA file (fd=4) in smaller chunks (usually 4KB). FASTA file is basically an index file for BAM file.

[1] system call logs for one of `DV`'s tasks [here](https://github.com/GoogleCloudPlatform/gcsfuse/files/1771987/task-7-sys-calls.log)

**Comment by nmousavi** [comment]
I have decreased the BAM read size to 1MB, and saw ~30x improvement (tasks finished in 69min). However, this is still ~5times slower compared with single worker/task scenario.

**Comment by jacobsa** [comment]
I'm not sure the `awaitRequestCancel` goroutines you quoted in the backtrace are surprising; that seems like it may be typical of any request in flight. (The `go-1.6` path component is surprising though. Are you building with an ancient version of Go?)

I do see there are two active calls to `Bucket.NewReader`:

    goroutine 107971 [select]:
    net/http.(*http2ClientConn).RoundTrip(0xc8a36cf760, 0xc8ffddf0a0, 0x0, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/h2_bundle.go:5180 +0xee5
    net/http.(*http2Transport).RoundTripOpt(0xc8200870e0, 0xc8ffddf0a0, 0xc8f297ac00, 0xc81d41a92c, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/h2_bundle.go:4812 +0x357
    net/http.(*http2Transport).RoundTrip(0xc8200870e0, 0xc8ffddf0a0, 0xffffffff, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/h2_bundle.go:4787 +0x41
    net/http.http2noDialH2RoundTripper.RoundTrip(0xc8200870e0, 0xc8ffddf0a0, 0x9baea8, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/h2_bundle.go:316 +0x3b
    net/http.(*Transport).RoundTrip(0xc8200e8000, 0xc8ffddf0a0, 0xc8ffddf0a0, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/transport.go:282 +0x352
    github.com/googlecloudplatform/gcsfuse/vendor/golang.org/x/oauth2.(*Transport).RoundTrip(0xc820079bc0, 0xc8ffddeee0, 0xc820079bc0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/golang.org/x/oauth2/transport.go:48 +0x234
    net/http.send(0xc8ffddeee0, 0x7f80b7c3d1c8, 0xc820079bc0, 0x0, 0x0, 0x0, 0xc82021d9a0, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/client.go:260 +0x6b7
    net/http.(*Client).send(0xc820079c20, 0xc8ffddeee0, 0x0, 0x0, 0x0, 0x1b, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/client.go:155 +0x185
    net/http.(*Client).doFollowingRedirects(0xc820079c20, 0xc8ffddeee0, 0xab1ac8, 0x0, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/client.go:475 +0x8a4
    net/http.(*Client).Do(0xc820079c20, 0xc8ffddeee0, 0xc741e0, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/client.go:188 +0xff
    github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs.(*bucket).NewReader(0xc820071f00, 0x7f80b7c76df8, 0xc8f1ffa700, 0xc88ffa3320, 0x0, 0x0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/read.go:88 +0xac2
    github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs.(*debugBucket).NewReader(0xc8200e1a20, 0x7f80b7c76df8, 0xc8f1ffa700, 0xc88ffa3320, 0x0, 0x0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/debug_bucket.go:136 +0x241
    github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/gcscaching.(*fastStatBucket).NewReader(0xc820336200, 0x7f80b7c76df8, 0xc8f1ffa700, 0xc88ffa3320, 0x0, 0x0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/gcscaching/fast_stat_bucket.go:125 +0x81
    github.com/googlecloudplatform/gcsfuse/internal/gcsx.(*contentTypeBucket).NewReader(0xc8204b4be0, 0x7f80b7c76df8, 0xc8f1ffa700, 0xc88ffa3320, 0x0, 0x0, 0x0, 0x0)
            <autogenerated>:6 +0x92
    github.com/googlecloudplatform/gcsfuse/internal/gcsx.(*randomReader).startRead(0xc82241e400, 0x124dde000, 0x20000, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/gcsx/random_reader.go:280 +0x245
    github.com/googlecloudplatform/gcsfuse/internal/gcsx.(*randomReader).ReadAt(0xc82241e400, 0x7f80b7bb5b78, 0xc8f177c930, 0xc83a292018, 0x20000, 0x20000, 0x124dde000, 0x0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/gcsx/random_reader.go:124 +0x113
    github.com/googlecloudplatform/gcsfuse/internal/fs/handle.(*FileHandle).Read(0xc822412a80, 0x7f80b7bb5b78, 0xc8f177c930, 0xc83a292018, 0x20000, 0x20000, 0x124dde000, 0x0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/fs/handle/file.go:110 +0x51b
    github.com/googlecloudplatform/gcsfuse/internal/fs.(*fileSystem).ReadFile(0xc82032c000, 0x7f80b7bb5b78, 0xc8f177c930, 0xc8e50b5a00, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/fs/fs.go:1561 +0x45e
    github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).handleOp(0xc8201ae960, 0xc8202fa000, 0x7f80b7bb5b78, 0xc8f177c930, 0x80c480, 0xc8e50b5a00)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil/file_system.go:177 +0xad5
    created by github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).ServeOps
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil/file_system.go:112 +0x18f
    
    goroutine 107972 [select]:
    net/http.(*http2ClientConn).RoundTrip(0xc8a36cf760, 0xc8ec4c9880, 0x0, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/h2_bundle.go:5180 +0xee5
    net/http.(*http2Transport).RoundTripOpt(0xc8200870e0, 0xc8ec4c9880, 0xc82045ec00, 0xc81a8a76d3, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/h2_bundle.go:4812 +0x357
    net/http.(*http2Transport).RoundTrip(0xc8200870e0, 0xc8ec4c9880, 0xffffffff, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/h2_bundle.go:4787 +0x41
    net/http.http2noDialH2RoundTripper.RoundTrip(0xc8200870e0, 0xc8ec4c9880, 0x9baea8, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/h2_bundle.go:316 +0x3b
    net/http.(*Transport).RoundTrip(0xc8200e8000, 0xc8ec4c9880, 0xc8ec4c9880, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/transport.go:282 +0x352
    github.com/googlecloudplatform/gcsfuse/vendor/golang.org/x/oauth2.(*Transport).RoundTrip(0xc820079bc0, 0xc8ec4c97a0, 0xc820079bc0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/golang.org/x/oauth2/transport.go:48 +0x234
    net/http.send(0xc8ec4c97a0, 0x7f80b7c3d1c8, 0xc820079bc0, 0x0, 0x0, 0x0, 0xc8201bb360, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/client.go:260 +0x6b7
    net/http.(*Client).send(0xc820079c20, 0xc8ec4c97a0, 0x0, 0x0, 0x0, 0x1b, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/client.go:155 +0x185
    net/http.(*Client).doFollowingRedirects(0xc820079c20, 0xc8ec4c97a0, 0xab1ac8, 0x0, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/client.go:475 +0x8a4
    net/http.(*Client).Do(0xc820079c20, 0xc8ec4c97a0, 0xc741e0, 0x0, 0x0)
            /usr/lib/go-1.6/src/net/http/client.go:188 +0xff
    github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs.(*bucket).NewReader(0xc820071f00, 0x7f80b7c76df8, 0xc8b9313d40, 0xc900468500, 0x0, 0x0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/read.go:88 +0xac2
    github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs.(*debugBucket).NewReader(0xc8200e1a20, 0x7f80b7c76df8, 0xc8b9313d40, 0xc900468500, 0x0, 0x0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/debug_bucket.go:136 +0x241
    github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/gcscaching.(*fastStatBucket).NewReader(0xc820336200, 0x7f80b7c76df8, 0xc8b9313d40, 0xc900468500, 0x0, 0x0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/gcscaching/fast_stat_bucket.go:125 +0x81
    github.com/googlecloudplatform/gcsfuse/internal/gcsx.(*contentTypeBucket).NewReader(0xc8204b4be0, 0x7f80b7c76df8, 0xc8b9313d40, 0xc900468500, 0x0, 0x0, 0x0, 0x0)
            <autogenerated>:6 +0x92
    github.com/googlecloudplatform/gcsfuse/internal/gcsx.(*randomReader).startRead(0xc8202a8380, 0x124dfe000, 0x20000, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/gcsx/random_reader.go:280 +0x245
    github.com/googlecloudplatform/gcsfuse/internal/gcsx.(*randomReader).ReadAt(0xc8202a8380, 0x7f80b7bb5b78, 0xc8f177c960, 0xc92e482018, 0x20000, 0x20000, 0x124dfe000, 0x0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/gcsx/random_reader.go:124 +0x113
    github.com/googlecloudplatform/gcsfuse/internal/fs/handle.(*FileHandle).Read(0xc820337e00, 0x7f80b7bb5b78, 0xc8f177c960, 0xc92e482018, 0x20000, 0x20000, 0x124dfe000, 0x0, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/fs/handle/file.go:110 +0x51b
    github.com/googlecloudplatform/gcsfuse/internal/fs.(*fileSystem).ReadFile(0xc82032c000, 0x7f80b7bb5b78, 0xc8f177c960, 0xc8e50b5b80, 0x0, 0x0)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/fs/fs.go:1561 +0x45e
    github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).handleOp(0xc8201ae960, 0xc8202fa000, 0x7f80b7bb5b78, 0xc8f177c960, 0x80c480, 0xc8e50b5b80)
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil/file_system.go:177 +0xad5
    created by github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).ServeOps
            /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil/file_system.go:112 +0x18f

It's not clear to me why these would be stuck here waiting for the start of the HTTP response, rather than in a call to `io.Reader.Read` consuming its body. Perhaps your stack traces are relevant after all, since they look like they're looking for an HTTP header. Could you build with the newest version of Go and try again?

Regarding the [later comment](https://github.com/GoogleCloudPlatform/gcsfuse/issues/262#issuecomment-369648391): how did you measure "doesn't help"? You may be wasting 7/8 of your GCS bandwidth on overreading, even if there's not an effect on timing.

**Comment by nmousavi** [comment]
New log can be found here: [gcsfuse-go-1-10-sigquit.log](https://github.com/GoogleCloudPlatform/gcsfuse/files/1805161/gcsfuse-go-1-10-sigquit.log)


re measurement: I haven't measured the total runtime as it takes >30h for the job to finish in this setup, but by just looking at the processed record, I could conclude that decreasing minReadSize has no impact. The reason for it can be that DeepVariant's read access is basically few small reads (1-15kB) from file A followed by a large read (128MB) from file B where most of the time is spent (2-3min every time). So improvement due to minReadSize has no impact on the overall runtime.

**Comment by jacobsa** [comment]
You do indeed still get stuck in `NewReader`:

```goroutine 23437 [runnable]:
net/http.(*http2ClientConn).roundTrip(0xc420352e00, 0xc4202c2500, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/http/h2_bundle.go:7442 +0x844
net/http.(*http2Transport).RoundTripOpt(0xc4201d40a0, 0xc4202c2500, 0xffffffffffffff00, 0x0, 0x0, 0xc420e2a348)
        /usr/local/go/src/net/http/h2_bundle.go:6929 +0x150
net/http.(*http2Transport).RoundTrip(0xc4201d40a0, 0xc4202c2500, 0xc420e2a378, 0x0, 0xc420e2a2e8)
        /usr/local/go/src/net/http/h2_bundle.go:6891 +0x3a
net/http.http2noDialH2RoundTripper.RoundTrip(0xc4201d40a0, 0xc4202c2500, 0x857525, 0x5, 0xc4201d21a8)
        /usr/local/go/src/net/http/h2_bundle.go:991 +0x39
net/http.(*Transport).RoundTrip(0xa9d0c0, 0xc4202c2500, 0xc4202c2500, 0x0, 0xffffffffffffffff)
        /usr/local/go/src/net/http/transport.go:380 +0xc36
github.com/googlecloudplatform/gcsfuse/vendor/golang.org/x/oauth2.(*Transport).RoundTrip(0xc420204090, 0xc4202c2400, 0xc420204090, 0x0, 0x0)
        /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/golang.org/x/oauth2/transport.go:48 +0xcf
net/http.send(0xc4202c2400, 0x8a6f40, 0xc420204090, 0x0, 0x0, 0x0, 0xc42076a2f0, 0x0, 0xc4208495e8, 0x1)
        /usr/local/go/src/net/http/client.go:252 +0x185
net/http.(*Client).send(0xc4202040f0, 0xc4202c2400, 0x0, 0x0, 0x0, 0xc42076a2f0, 0x0, 0x1, 0x10)
        /usr/local/go/src/net/http/client.go:176 +0xfa
net/http.(*Client).Do(0xc4202040f0, 0xc4202c2400, 0xaa2f00, 0xc420447b20, 0x19)
        /usr/local/go/src/net/http/client.go:615 +0x28d
github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs.(*bucket).NewReader(0xc42020a0c0, 0x8aaa00, 0xc42020bc80, 0xc4203c13a0, 0x0, 0x0, 0x0, 0x0)
        /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/read.go:88 +0x526
github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs.(*debugBucket).NewReader(0xc4202000c0, 0x8aaa00, 0xc42020bc80, 0xc4203c13a0, 0xc4204a1000, 0xc420022088, 0x0, 0x0)
        /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/debug_bucket.go:136 +0x16f
github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/gcscaching.(*fastStatBucket).NewReader(0xc42020a400, 0x8aaa00, 0xc42020bc80, 0xc4203c13a0, 0x10, 0x805980, 0xc42020bc01, 0xc4201560a0)
        /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/gcloud/gcs/gcscaching/fast_stat_bucket.go:125 +0x52
github.com/googlecloudplatform/gcsfuse/internal/gcsx.(*randomReader).startRead(0xc420ed5600, 0x37321000, 0x20000, 0x0, 0x0)
        /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/gcsx/random_reader.go:271 +0x2e7
github.com/googlecloudplatform/gcsfuse/internal/gcsx.(*randomReader).ReadAt(0xc420ed5600, 0x8aaac0, 0xc42213c840, 0xc4206bc018, 0x20000, 0x20000, 0x37321000, 0x0, 0x0, 0x0)
        /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/gcsx/random_reader.go:124 +0x628
github.com/googlecloudplatform/gcsfuse/internal/fs/handle.(*FileHandle).Read(0xc420ed5200, 0x8aaac0, 0xc42213c840, 0xc4206bc018, 0x20000, 0x20000, 0x37321000, 0x0, 0x0, 0x0)
        /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/fs/handle/file.go:110 +0x2ba
github.com/googlecloudplatform/gcsfuse/internal/fs.(*fileSystem).ReadFile(0xc42046a0c0, 0x8aaac0, 0xc42213c840, 0xc42018e640, 0x0, 0x0)
        /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/internal/fs/fs.go:1561 +0x2eb
github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).handleOp(0xc4203a4620, 0xc42046a180, 0x8aaac0, 0xc42213c840, 0x7a44a0, 0xc42018e640)
        /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil/file_system.go:177 +0xa89
created by github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).ServeOps
        /home/nmousavi/go/src/github.com/googlecloudplatform/gcsfuse/vendor/github.com/jacobsa/fuse/fuseutil/file_system.go:112 +0xca
```

I still don't understand why that is. I guess it would help to know how much data is flowing through the HTTP connection -- are we receiving the entire body in the call to `http.Client.Do`? The code assumes that we don't, but rather just receive a reader for the body, which is received asynchronously.

Still, this call [doesn't happen](https://github.com/GoogleCloudPlatform/gcsfuse/blob/25c2c1b11b9dbc34920cbf5a4da236ab2d9a0341/internal/fs/handle/file.go#L102-L104) under the inode lock, so it's not clear to me why this should matter. I suppose maybe it could be preventing you from multiplexing on a shared HTTP connection, blocking progress of other requests while the body is returned?

**Comment by nmousavi** [comment]
Not sure what would this (checking if the entire body is received) reveal? We always read async, no? What can go wrong here? Do goroutines block each other when reading from responses?

This is probably the case that shared connection is not multiplexed. But where is the bottleneck, and how they block each other? How can we prove this is the case (any more log)? Is it possible to spin off multiple connections?

To recall, if we run multiple gcsfuse, or if reduce the read range (from 128MB to 128KB) the issue disappears. I don't understand how/why?

**Comment by jacobsa** [comment]
If the HTTP library is consuming 128 MiB from the server "atomically" in the sense that it's not interleaving other responses, then that's clearly what's blocking other requests. I don't know why that would happen, but HTTP/2 is complicated and perhaps the server is setting some flow control options that make this necessary? I can't help you with how to figure that out, but I guess logging in `net/http` and/or use of its debugging environment variables would help. You can probably use `--debug_http` with gcsfuse or similar to get a request that you can use to reproduce this without needing gcsfuse in the mix.

**Comment by nmousavi** [comment]
The issue exists even with HTTP/2 off.   I use env var to turn off HTTP/2

GODEBUG=http2client=0  # disable HTTP/2 client support

**Comment by jacobsa** [comment]
This thread seems to have dried up; where are you with the debugging here? It seems like this is a problem with the HTTP library if anything, or do we have evidence that gcsfuse is doing something wrong?

**Comment by nmousavi** [comment]
I'm not convinced that http flow control is the culprit, or there is no way gcsfuse coulf avoid it. I will update this issue once I come up with a test to prove or disprove it.

**Comment by iskandr** [comment]
I'm having a similar problem running bioinformatics tools on whole genome sequencing data. Maybe either a disclaimer that gcsfuse isn't suitable for these minds of workloads *or* a blog post showing how to do it right would be helpful.

**Comment by jacobsa** [comment]
I'm happy to update the existing [performance](https://github.com/googlecloudplatform/gcsfuse#performance) section in the README if you can characterize the workload that's causing you trouble. As far as I can tell the issue in this thread is small random reads. Is that what you're concerned about too?

**Comment by jacobsa** [comment]
Marking closed. Please open a new issue or send a pull request if you've got suggestions about how documentation should be changed.

**Comment by bjornleffler** [comment]
I can reproduce this.

**Comment by cbandy** [comment]
Release notes for Go 1.12 mention that earlier versions "wait for requests to finish instead of creating new connections." Related?

https://golang.org/doc/go1.12#net/http

**Comment by bjornleffler** [comment]
I have found one major issue, where multiple sequential readers of the same files causes read performance to drop, especially for large files. Working on it.

**Comment by bjornleffler** [comment]
If you were affected by this issue, please test version 0.27.0, which has improvement for parallel / multi threaded access.

**Comment by gigadeplex** [comment]
This is still horrible, I just ran into this problem as well, I'm using 0.28.1

**Comment by nicklasring** [comment]
> Looking into system call logs, I see `DeepVariant` (`DV`) makes a very large (exactly 128MB!?!) read request to `gcsfuse` every so often, and this is when everything freezes for minutes (1-3min); `DV` waits for read() sys call, and gcsfuse waits for some lock.
> 
> system call log for DV 
> 
> 19:06:23 lseek(6, 18099034291, SEEK_SET) = 18099034291
> 19:06:23 read(6"\37\213\10\4\0\0\0\0\0\377\6\0BC\2\0_L\315\275\t\230$\371U\37\3303\243\356\254\256"..., 134217728) = 134217728
> 19:08:24 lseek(4, 235593595, SEEK_SET)  = 235593595
> 19:08:24 read(4, "\37\213\10\4\0\0\0\0\0\377\6\0BC\2\0uM\235\235[\226\344J\216k\3775+.\373\340"..., 4096) = 4096
> system call log for gcsfuse 
> 
> 19:06:25 futex(0xc820027508, FUTEX_WAKE, 1) = 1
> 19:06:25 futex(0xc75128, FUTEX_WAIT, 0, NULL) = 0 
> 19:06:26 futex(0xc75128, FUTEX_WAIT, 0, NULL) = 0
> 19:06:27 futex(0xc75128, FUTEX_WAIT, 0, NULL) = 0 
> 19:07:30 futex(0xc820027908, FUTEX_WAKE, 1) = 1     # 1 min lock wait!
> 19:07:30 read(7, "\2533d8\231@,\22<#\1\332\231\31\207\325\333x\372k\331d\26\337t'\f\373y\207V\30"..., 8230) = 8230
> 19:07:30 sched_yield()                  = 0
> 
> We can see gcsfuse spends most of its time (13%) on kernel lock `__lock_text_start`.
> 
> <img alt="screen shot 2018-02-28 at 10 31 00 pm" width="1440" src="https://user-images.githubusercontent.com/4221028/36825789-5123c33a-1cd7-11e8-8b10-83440469c1d5.png">
> 
> From Network dump, I see more Application Data packet and relatively slow Acks are sent to `GCS` during this time. For example, only 454MB was acked between [19:36:07, 19:37:22] which is only 7MB/s. No TCP zeroWindow Err in this particular period, but this is something that happens for `gcsfuse` (not very frequent though).
> 
> ```
> 11	19:36:07.460602	10.142.0.8	74.125.141.95	TCP	66	34908 → 443 [ACK] Seq=67 Ack=40150 Win=24323 Len=0 TSval=609828941 TSecr=3815364206
> ...
> 83526	19:37:22.274651	10.142.0.8	74.125.141.95	TCP	66	34908 → 443 [ACK] Seq=75858 Ack=476136771 Win=24020 Len=0 TSval=609903749 TSecr=3815439013
> ```
> 
> TCP zero window
> 
> <img alt="screen shot 2018-02-28 at 11 01 59 am" width="1440" src="https://user-images.githubusercontent.com/4221028/36848116-4e5c9634-1d2e-11e8-888a-63f1c7ac4807.png">
> 
> I sent a `SIGQUIT` to `gcsfuse` when it was stalled on the lock:
> 
> There are many awaitRequestCancel thread with O(min) wait
> 
> ```
> goroutine 91265 [select, 4 minutes]:
> net/http.(*http2clientStream).awaitRequestCancel(0xc8e1b9c3c0, 0xc8223e68a0)
> 	/usr/lib/go-1.6/src/net/http/h2_bundle.go:4731 +0x156
> created by net/http.(*http2clientConnReadLoop).processHeaderBlockFragment
> 	/usr/lib/go-1.6/src/net/http/h2_bundle.go:5731 +0x7c9
> 
> goroutine 92709 [select, 4 minutes]:
> net/http.(*http2clientStream).awaitRequestCancel(0xc84680d040, 0xc9bc313da0)
> 	/usr/lib/go-1.6/src/net/http/h2_bundle.go:4731 +0x156
> created by net/http.(*http2clientConnReadLoop).processHeaderBlockFragment
> 	/usr/lib/go-1.6/src/net/http/h2_bundle.go:5731 +0x7c9
> ...
> ```
> 
> [1] network dump [here](https://nmousavi.users.x20web.corp.google.com/tcpdump-gcsfuse.pcap) (~1GB) [2] `gcsfuse` log with `SIGQUIT` [here](https://nmousavi.users.x20web.corp.google.com/gcsfuse-sigquit.log) (36MB)

Also affected by this, found this issue by searching for the timestamp at 06:25 , every day we get a spike in requests seemingly caused by gcsfuse.
